File: app/frontend/src/core/utils/__tests__/logger.test.ts
import { logger } from '../logger';

describe('logger', () => {
  let consoleLogSpy: jest.SpyInstance;
  let consoleErrorSpy: jest.SpyInstance;
  let consoleWarnSpy: jest.SpyInstance;
  let consoleDebugSpy: jest.SpyInstance;

  beforeEach(() => {
    consoleLogSpy = jest.spyOn(console, 'log').mockImplementation();
    consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation();
    consoleWarnSpy = jest.spyOn(console, 'warn').mockImplementation();
    consoleDebugSpy = jest.spyOn(console, 'debug').mockImplementation();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('info', () => {
    it('should log info message with timestamp', () => {
      logger.info('test message');
      expect(consoleLogSpy).toHaveBeenCalledWith(expect.stringMatching(/^\[\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z\] INFO: test message$/));
    });

    it('should include metadata in log message', () => {
      logger.info('test message', { key: 'value' });
      expect(consoleLogSpy).toHaveBeenCalledWith(expect.stringMatching(/^\[\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z\] INFO: test message \| {"timestamp":".+","level":"info","key":"value"}$/));
    });
  });

  describe('error', () => {
    it('should log error message with timestamp', () => {
      logger.error('test error');
      expect(consoleErrorSpy).toHaveBeenCalledWith(expect.stringMatching(/^\[\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z\] ERROR: test error$/));
    });

    it('should include error details in log message', () => {
      const error = new Error('test error');
      logger.error('error occurred', error);
      expect(consoleErrorSpy).toHaveBeenCalledWith(expect.stringMatching(/^\[\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z\] ERROR: error occurred \| {"timestamp":".+","level":"error","error":{"name":"Error","message":"test error","stack":".+"}}$/));
    });

    it('should include metadata and error details', () => {
      const error = new Error('test error');
      logger.error('error occurred', error, { context: 'test' });
      expect(consoleErrorSpy).toHaveBeenCalledWith(expect.stringMatching(/^\[\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z\] ERROR: error occurred \| {"timestamp":".+","level":"error","context":"test","error":{"name":"Error","message":"test error","stack":".+"}}$/));
    });
  });

  describe('warn', () => {
    it('should log warning message with timestamp', () => {
      logger.warn('test warning');
      expect(consoleWarnSpy).toHaveBeenCalledWith(expect.stringMatching(/^\[\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z\] WARN: test warning$/));
    });

    it('should include metadata in log message', () => {
      logger.warn('test warning', { key: 'value' });
      expect(consoleWarnSpy).toHaveBeenCalledWith(expect.stringMatching(/^\[\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z\] WARN: test warning \| {"timestamp":".+","level":"warn","key":"value"}$/));
    });
  });

  describe('debug', () => {
    const originalDebug = process.env.DEBUG;

    beforeEach(() => {
      process.env.DEBUG = 'true';
    });

    afterEach(() => {
      process.env.DEBUG = originalDebug;
    });

    it('should log debug message when DEBUG is true', () => {
      logger.debug('test debug');
      expect(consoleDebugSpy).toHaveBeenCalledWith(expect.stringMatching(/^\[\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z\] DEBUG: test debug$/));
    });

    it('should include metadata in debug message', () => {
      logger.debug('test debug', { key: 'value' });
      expect(consoleDebugSpy).toHaveBeenCalledWith(expect.stringMatching(/^\[\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z\] DEBUG: test debug \| {"timestamp":".+","level":"debug","key":"value"}$/));
    });

    it('should not log debug message when DEBUG is false', () => {
      process.env.DEBUG = 'false';
      logger.debug('test debug');
      expect(consoleDebugSpy).not.toHaveBeenCalled();
    });
  });
}); 

File: backend/app/tests/typescript/utils/logger.test.py
from typing import Any


describe('Logger', () => {
  let consoleSpy: jest.SpyInstance
  let logger: Logger
  beforeEach(() => {
    logger = new Logger()
    consoleSpy = jest.spyOn(console, 'debug').mockImplementation()
    jest.spyOn(console, 'info').mockImplementation()
    jest.spyOn(console, 'warn').mockImplementation()
    jest.spyOn(console, 'error').mockImplementation()
  })
  afterEach(() => {
    jest.restoreAllMocks()
  })
  describe('Log Level Filtering', () => {
    it('should log all levels when minLevel is DEBUG', () => {
      logger.configure({ minLevel: LogLevel.DEBUG })
      logger.debug('debug message')
      logger.info('info message')
      logger.warn('warn message')
      logger.error('error message')
      expect(console.debug).toHaveBeenCalled()
      expect(console.info).toHaveBeenCalled()
      expect(console.warn).toHaveBeenCalled()
      expect(console.error).toHaveBeenCalled()
    })
    it('should only log ERROR when minLevel is ERROR', () => {
      logger.configure({ minLevel: LogLevel.ERROR })
      logger.debug('debug message')
      logger.info('info message')
      logger.warn('warn message')
      logger.error('error message')
      expect(console.debug).not.toHaveBeenCalled()
      expect(console.info).not.toHaveBeenCalled()
      expect(console.warn).not.toHaveBeenCalled()
      expect(console.error).toHaveBeenCalled()
    })
  })
  describe('Message Formatting', () => {
    it('should include timestamp when enabled', () => {
      logger.configure({ timestamp: true })
      logger.info('test message')
      const call = (console.info as jest.Mock).mock.calls[0][0]
      expect(call).toMatch(/\[\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\.\d{3}\]/)
    })
    it('should not include timestamp when disabled', () => {
      logger.configure({ timestamp: false })
      logger.info('test message')
      const call = (console.info as jest.Mock).mock.calls[0][0]
      expect(call).not.toMatch(/\[\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\.\d{3}\]/)
    })
    it('should include prefix when set', () => {
      const prefix = 'TEST'
      logger.configure({ prefix })
      logger.info('test message')
      const call = (console.info as jest.Mock).mock.calls[0][0]
      expect(call).toContain(`[${prefix}]`)
    })
    it('should handle additional arguments', () => {
      const additionalArg = { key: 'value' }
      logger.info('test message', additionalArg)
      expect(console.info).toHaveBeenCalledWith(
        expect.any(String),
        additionalArg
      )
    })
  })
  describe('Configuration', () => {
    it('should use default configuration when none provided', () => {
      const defaultLogger = new Logger()
      defaultLogger.debug('test')
      const call = (console.debug as jest.Mock).mock.calls[0][0]
      expect(call).toMatch(/\[\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\.\d{3}\]/)
    })
    it('should update configuration with configure method', () => {
      logger.configure({
        minLevel: LogLevel.WARN,
        timestamp: false,
        prefix: 'NEW'
      })
      logger.warn('test message')
      const call = (console.warn as jest.Mock).mock.calls[0][0]
      expect(call).not.toMatch(/\[\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\.\d{3}\]/)
      expect(call).toContain('[NEW]')
    })
  })
}) 